rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,209 @@
+However, we can easily extend the search space to include more complex expressions with the cost of an exponential increase of the running time of the synthesizer.
+2. the synthesizer suffers from the same limitation as the loop contract verification in CBMC. For example, it does not support unbounded quantifiers, or dynamic allocations in the loop body. 
+
+## User Experience
+
+Once this RFC has been stabilized, users will be able to use the new command-line flag `--synthesize-loop-contracts` to run the synthesizer, which will attempt to synthesize loop contracts, and verify programs with the synthesized loop contracts.
+
+
+#### Limit Resource Used by Synthesizer for Termination
+Without a resource limit, an enumerative synthesizer may run forever to exhaust a search space consisting of an infinite number of candidates, especially when there is no solution in the search space.
+So, for the guarantee of termination, we provide users options: `--limit-synthesis-time T` to limit the running time of the synthesizer.
+
+
+#### Output of Kani when the Synthesizer is Enabled
+When the flag `--synthesize-loop-contracts` is provided, Kani will report different result for different cases
+1. When there exists some loop invariant in the candidate space with which all assertions can be proved, Kani will synthesize the loop contracts, verify the program with the synthesized loop contracts, and report verification SUCCESS;
+2. When no working candidate has been found in the search space within the specified limits, Kani will report the verification result with the best-effort-synthesized loop contracts.
+Note that as loop contracts are over-approximations of the loop, the violated assertions in this case may be spurious.
+So we will report the violated assertions as `UNDETERMINED` instead of `FAILED`.
+
+A question about how do we print the synthesized loop contracts when users request is discussed in **Open question**.
+
+## Detailed Design
+The synthesizer ```goto-synthesizer``` is implemented in the repository of `CBMC`, takes as input a goto binary, and outputs a new goto binary with the synthesized loop contracts applied.
+Currently, Kani invokes `goto-instrument` to instrument the goto binary `main.goto` into a new goto binary `main_instrumented.goto`, and then invokes ```cbmc``` on `main_instrumented.goto` to get the verification result.
+The synthesis will happen between calling `goto-instrument` and calling ```cbmc```.
+That is, we invoke ```goto-synthesizer``` on ```main_instrumented.goto``` to produce a new goto binary ```main_synthesized.goto```, and then call ```cbmc``` on `main_synthesized.goto` instead. 
+
+When invoking ```goto-synthesizer```, we pass the following parameters to it with the flags built in `goto-synthesizer`:
+- the resource limit of the synthesis;
+- the solver options to specify what SAT solver we use to verify invariant candidates.
```

Comment:
Is this still true? I recall you changed it so the SAT solver is just the one used for verification.

Reply:
Yes, we propagate the solver options from Kani to goto-synthesizer.